src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netstandard.cs

```diff
@@ -44,6 +46,21 @@ private static float CosineSimilarityCore(ReadOnlySpan<float> x, ReadOnlySpan<fl
                 }
                 while (i <= oneVectorFromEnd);
 
+                // Process the last vector in the span, masking off elements already processed.
+                if (i != x.Length)
+                {
+                    Vector<float> xVec = AsVector(ref xRef, x.Length - Vector<float>.Count);
+                    Vector<float> yVec = AsVector(ref yRef, x.Length - Vector<float>.Count);
+
+                    Vector<float> remainderMask = LoadRemainderMaskSingleVector(x.Length - i);
```

Comment:
interesting trick with the mask, do we want to have something like this as a public API in vector? (presumably, it can be accelerated on avx512 with masks)

Reply:
For handling remainders at the end, and for helping with alignment at the beginning, I think it'd be nice if we added APIs like:
```diff
struct Vector128<T>
{
+    public static Vector128<T> CreateStartMask(int length); // e.g. for T == ushort, length == 2, it'd produce [0xFFFF, 0xFFFF, 0, 0, 0, 0, 0, 0]
+    public static Vector128<T> CreateEndMask(int length); // e.g. for T == ushort, length == 2, it'd produce [0, 0, 0, 0, 0, 0, 0xFFFF, 0xFFFF]
}
```
The table I've included here would need to be adapted to support all relevant size Ts, and given how big it would be (back of the napkin, 4K?), and how annoying it is to produce :smile:, it'd be nice if there were a built-in mechanism for it... especially if it wouldn't be needed at all with AVX512.  I tried to come up with some other scheme for achieving this, but couldn't come up with anything better.

Reply:
I was thinking of maybe something like this:
```C#
Vector128<float> mask = Vector128.ConditionalSelect(
    Vector128.GreaterThan(Vector128.Create(0, 1, 2, 3), Vector128.Create(3 - length)).AsSingle(),
    Vector128<float>.AllBitsSet,
    Vector128<float>.Zero);
```

Reply:
Ah, right, `ShiftLeftLogical128BitLane` won't work -- it requires a constant argument or will be jitted to a giant switch with questionable perf 🙁 `ConditionalSelect` looks interesting

Reply:
> ConditionalSelect looks interesting

It's just not quite as good, e.g.
```C#
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Runtime.Intrinsics;

internal class Program
{
    private static void Main()
    {
        for (int i = 0; i <= 5; i++)
        {
            Console.WriteLine(CreateRemainder1(i));
            Console.WriteLine(CreateRemainder2(i));
            Console.WriteLine();
        }
    }

    static Vector128<float> CreateRemainder1(int length)
    {
        return Vector128.ConditionalSelect(
            Vector128.GreaterThan(Vector128.Create(0, 1, 2, 3), Vector128.Create(3 - length)).AsSingle(),
            Vector128<float>.AllBitsSet,
            Vector128<float>.Zero);
    }

    private static unsafe Vector128<float> CreateRemainder2(int validItems) =>
        Vector128.LoadUnsafe(
            ref Unsafe.As<uint, float>(ref MemoryMarshal.GetReference(RemainderUInt32Mask_16x16)),
            (uint)((validItems * 16) + 12)); // last four floats in the row

    private static ReadOnlySpan<uint> RemainderUInt32Mask_16x16 => new uint[]
    {
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0x00000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
        0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
    };
}
```

yields:
```asm
; Method Program:CreateRemainder1(int):System.Runtime.Intrinsics.Vector128`1[float] (FullOpts)
G_M000_IG01:                ;; offset=0x0000
       vzeroupper 

G_M000_IG02:                ;; offset=0x0003
       mov      eax, edx
       neg      eax
       add      eax, 3
       vmovd    xmm0, eax
       vpbroadcastd xmm0, xmm0
       vmovups  xmm1, xmmword ptr [reloc @RWD00]
       vpcmpgtd xmm0, xmm1, xmm0
       vpcmpeqd xmm1, xmm1, xmm1
       vxorps   xmm2, xmm2, xmm2
       vblendvps xmm0 xmm2, xmm1, xmm0
       vmovups  xmmword ptr [rcx], xmm0
       mov      rax, rcx

G_M000_IG03:                ;; offset=0x0034
       ret      
RWD00  	dq	0000000100000000h, 0000000300000002h
; Total bytes of code: 53
```
and
```asm
; Method Program:CreateRemainder2(int):System.Runtime.Intrinsics.Vector128`1[float] (FullOpts)
G_M000_IG01:                ;; offset=0x0000
       vzeroupper 

G_M000_IG02:                ;; offset=0x0003
       shl      edx, 4
       add      edx, 12
       mov      eax, edx
       mov      rdx, 0x12312192CD8
       vmovups  xmm0, xmmword ptr [rdx+4*rax]
       vmovups  xmmword ptr [rcx], xmm0
       mov      rax, rcx

G_M000_IG03:                ;; offset=0x0021
       ret      
; Total bytes of code: 34
```

Reply:
Although, in practice, we're talking about a few instructions per larger operation.  I'd dismissed it because it's not quite as fast, but it's so close, probably better to avoid the large table.  I'll change it in a subsequent PR.